sdk/communication/AzureCommunication/Source/Authentication/CommunicationTokenRefreshOptions.swift

```diff
@@ -0,0 +1,58 @@
+// --------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// The MIT License (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the ""Software""), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+import Foundation
+
+/**
+ The Communication Token Refresh Options. Used to initialize a `CommunicationUserCredential`
+ - SeeAlso: ` CommunicationUserCredential.init(...)`
+*/
+@objcMembers public class CommunicationTokenRefreshOptions: NSObject {
+    var initialToken: String?
+    var refreshProactively: Bool
+    var tokenRefresher: (@escaping TokenRefreshOnCompletion) -> Void
+    /**
+     Initializes a new instance of `CommunicationTokenRefreshOptions`
+     The cached token is updated if `token(completionHandler: )` is called and if the difference between the current time
+     and token expiry time is less than 120s.
+     If `refreshProactively` parameter  is `true`:
+        - The cached token will be updated in the background when the difference between the current time and token expiry time is less than 600s.
+        - The cached token will be updated immediately when the constructor is invoked and `initialToken` is expired
+     - Parameters:
+        - initialToken: The initial value of the token.
+        - refreshProactively: Whether the token should be proactively refreshed in the background.
+        - tokenRefresher: Closure to call when a new token value is needed.
+     */
+    public init(
+        initialToken: String? = nil,
+        refreshProactively: Bool = false,
+        tokenRefresher: @escaping (@escaping TokenRefreshOnCompletion) -> Void
```

Comment:
This is not a syntax I've seen before. Why is the double @escaping needed?

Reply:
I agree this looks a little weird to read. 
`TokenRefreshOnCompletion` is a typealias which is closure: `public typealias TokenRefreshOnCompletion = (String?, Error?) -> Void` 
So `(@escaping TokenRefreshOnCompletion) -> Void` the inner @escaping refers to escaping inner closure. 
And the outer @escaping is escaping the closure `(@escaping TokenRefreshOnCompletion) -> Void`

Reply:
Should I typealias `(@escaping TokenRefreshOnCompletion) -> Void` to SomeTokenRefresher
So then it will look @escaping SomeTokenRefresher and hide it all?

Reply:
Hmmm... interesting. So the closure takes a closure. If the token refresh closure always needs to be escaping (which i presume it would) then I think it makes sense to move the inner escaping to the typealias---provided the compiler doesn't whine.